prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,535 @@
+/* SPDX-License-Identifier: BSD-2-Clause OR GPL-2.0-only */
+/* SPDX-FileCopyrightText: Copyright (c) 2013-2015 Intel Corporation, Inc.  All rights reserved. */
+/* SPDX-FileCopyrightText: Copyright Amazon.com, Inc. or its affiliates. All rights reserved. */
+
+#include <errno.h>
+#include <string.h>
+#include "config.h"
+#include <ofi_mem.h>
+#include "dgram/efa_dgram_ep.h"
+#include "efa.h"
+#include "efa_cq.h"
+#include "efa_av.h"
+#include <infiniband/verbs.h>
+
+
+static inline uint64_t efa_cq_opcode_to_fi_flags(enum ibv_wc_opcode opcode) {
+	switch (opcode) {
+	case IBV_WC_SEND:
+		return FI_SEND | FI_MSG;
+	case IBV_WC_RECV:
+		return FI_RECV | FI_MSG;
+	case IBV_WC_RDMA_WRITE:
+		return FI_RMA | FI_WRITE;
+	case IBV_WC_RECV_RDMA_WITH_IMM:
+		return FI_REMOTE_CQ_DATA | FI_RMA | FI_REMOTE_WRITE;
+	case IBV_WC_RDMA_READ:
+		return FI_RMA | FI_READ;
+	default:
+		assert(0);
+		return 0;
+	}
+}
+
+static void efa_cq_read_context_entry(struct ibv_cq_ex *ibv_cqx, void *buf)
```

Comment:
I am not sure we have to distinguish the format here. We can use an entry that can read all the formats, like the `data_entry`. We will not need tagged format here as efa device doesn't support tagged

Reply:
efa_rdm's ope used `tagged_entry` as `cq_entry` because it supports tagged

Reply:
Think about it again, using `tagged_entry` is good as you have used it in t he cq poll code, and if in the future efa device supports tagged, we don't need to change this format